Fixed #25588 -- Added spatial lookups to RasterField

yellowcap opened the pull request:
Created first version for lookup functionality. This works for the tests created, but I am not sure if it is a good implementation. I would like to get some feedback on the approach.

Some ideas for structuring of the band level lookup api:

- The default is always the first band, lookups can be done normally for the default
- On the rhs, band specific lookups could be similar to distance lookups, i.e. provide a tuple `(rast, band_nr)` or a triple for distance and relate lookups for a specific band `(rast, band_nr, D(mi=3))` and `(rast, band_nr, 'T*T***FF*')`
- For the lhs I am less sure about a good approach, one option is to add another double underscore part for the band nr, `rast__band_nr__dwithin`.

yellowcap commented:
The current lookup api works ok with rasters if using the first band. For passing band indices, the api has to be extended a bit, I added some thoughts about that in the comment above.

claudep commented:
Daniel, could you make a separate PR with getting/setting a raster srid, including docs?

claudep commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -5,32 +5,61 @@
 from django.contrib.gis.db.backends.postgis.adapter import PostGISAdapter
 from django.contrib.gis.db.backends.postgis.pgraster import (
-    from_pgraster, to_pgraster,
+    from_pgraster, get_pgraster_srid, to_pgraster,
```
See b55b34129b14

yellowcap commented:
Here is the PR for the GDALRaster `srid` property https://github.com/django/django/pull/5493

yellowcap commented:
Here is an alternative proposal for the lookup api, which I implemented in the current version of this PR.

The band index could be considered a rhs parameter, not sure if that makes sense conceptually.

But if that is acceptable, the band indices could both be passed on the rhs through a tuple. A lookup for band 3 on the lhs and band 2 on the rhs would be

    rast_field__intersects=(rast_input, (3,2))

and if a single band index is provided, it is applied to both sides, the following would use band 4 on both rasters:

    `rast_field__intersects=(rast_input, 4)`

For mixed lookups, the same syntax could work as well

    `rast_field__intersects=(geom_input, 1)`
    `geom_field__intersects=(rast_input, 1)`

would use the band 1 on the raster input on lhs or rhs.

yellowcap commented:
Is there anything in particular to run the widows tests? I can't figure out what is going wrong there, it does not seem to run any tests.

timgraham commented:
Looks like this is still waiting on a API design review? Asking for feedback on the django-developers mailing list might help it move forward.

claudep commented:
I'm sorry that my Django's time has reduced a lot recently :-(

yellowcap commented:
I added a second commit that alters the lhs band lookup syntax to be based on double  underscores `__`. 

The alteration is simple, mainly just adding a transform class to handle band index lookups.

Most of the compiling of band scpecific sql is currently done in the operations module. This could be moved further up, but I am not sure how standardized the raster sql is when compared to sqlite or oracle. So for now it sits in the `check_raster` method of the `PostGISOperator`.

timgraham commented:
As I said [on the mailing list](https://groups.google.com/d/topic/django-developers/ew6o7VoWRqw/discussion), I think the underscore is consistent with ArrayField's [index transforms](https://docs.djangoproject.com/en/1.9/ref/contrib/postgres/fields/#index-transforms) so that seems okay to me. I guess this is just missing docs then.

yellowcap commented:
Thanks @timgraham I will get back to this soon to add documentation for the lookups.

yellowcap commented:
I wrote documentation for the lookups trying to not add too much new parts.

The main challenge is that the raster lookups have more possible combinations. Such as

    rast__contains=rast, rast__1__contains=(rst, 2), rast__contains=(rst, 2), rast__1__contains=(rst, 2)

and for the SQL Equivalence snippets, the `ST_Polygon` gets injected depending on the case as well, so there are many possible lookup combinations for each lookup.

I am not sure how many of these to show explicitly in the examples in the reference. Currently the possible combinations are shown in a separate section at the top of the geoqueryset reference, and there are no new examples in the individual lookup references.

Please have a look when you have time and let me know if this setup makes sense and whether the flow is ok.

Then I still want to write a few more tests, but aside from that I think its ready (the current tests for the distance lookups are ok. For the other ones the tests just run all possible lookup combinations, but do not check if the output is the expected one. So that needs improvement).

timgraham commented on docs/ref/contrib/gis/db-api.txt:
```diff
@@ -52,8 +52,13 @@ on a different spatial backend.
 
-``RasterField`` is currently only implemented for the PostGIS backend. Spatial
-queries (such as lookups and distance) are not yet available for raster fields.
+.. versionchanged:: 1.10
```
These annotations generally go at the bottom of a section.

timgraham commented on docs/ref/contrib/gis/db-api.txt:
```diff
@@ -52,8 +52,13 @@ on a different spatial backend.
+
+    The ``RasterField`` now supports spatial lookups.
+
+The ``RasterField`` is currently only implemented for the PostGIS backend.
```
No need to add "The"

timgraham commented on docs/ref/contrib/gis/db-api.txt:
```diff
@@ -136,26 +141,79 @@ Spatial Lookups
+<spatial-lookups>`.
+
+Geometry Lookups
+----------------
```
Add blank line after a heading.

timgraham commented on docs/ref/contrib/gis/geoquerysets.txt:
```diff
@@ -40,12 +88,13 @@ MySQL       ``MBRContains(poly, geom)``
 SpatiaLite  ``MbrContains(poly, geom)``
 ==========  ==========================
 
+
```
revert

yellowcap commented:
Updated documentation based on comments. Also added additional logical tests for three lookups, so on my end this is good to go.

timgraham commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -17,24 +19,73 @@
+                template_params['lhs'] = 'ST_Polygon(%s)' % template_params['lhs']
+            if rhs_is_raster:
+                template_params['rhs'] = 'ST_Polygon(%s)' % template_params['rhs']
+        elif self.raster == 'bilateral':
```
maybe it would be a good idea to add a `BILATERAL = 'bilateral'` module constant to help prevent against typos of that "magic string"?

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -157,6 +160,81 @@ def get_placeholder(self, value, compiler, connection):
+                try:
+                    obj = GDALRaster(obj)
+                except GDALException:
+                    raise ValueError('Could not create spatial object from lookup value.')
```
include the value in these error messages for easier debugging?

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -82,9 +82,9 @@ def alter_gis_model(self, migration_class, model_name, field_name,
         operation = migration_class(*args)
         new_state = project_state.clone()
         operation.state_forwards('gis', new_state)
+        self.current_state = new_state
```
Did you run into problems that necessitated this change? It seems like it would cause incorrect tear down if `operation.database_forwards` fails (trying to rollback that operation that failed).

yellowcap replied:
Yes, these tests fail on spatialite with the following message:

    OperationalError: table "gis_neighborhood" already exists

Moving that line up fixes the problem. I remember it took me a while to even identify the problem when I was writing the lookups. I can try to look further into why this happens, but honestly I don't understand what is going on. Any hints/help regarding this are most welcome.

timgraham commented on tests/gis_tests/rasterapp/test_rasterfield.py:
```diff
@@ -1,20 +1,47 @@
     available_apps = ['gis_tests.rasterapp']
 
+    def setUp(self):
+        from django.contrib.gis.gdal import GDALRaster
```
How about this at the top of the file:
```python
try:
    from django.contrib.gis.gdal import GDALRaster
except ImportError: # or whatever exception it is
    GDALRaster = None
```
to avoid the duplicated inner imports?

timgraham commented:
Cosmetic edits at https://github.com/geodesign/django/pull/2. Don't forget a mention in the release notes as well.

@claudep, care to review this?

yellowcap commented:
For now I reverted the one-line change on the `gis_migrations.gis_operations` tests to replicate the spatialite error on jenkins.

I also added tests for the error messages on invalid lookup input. The error messages include the lookup value as suggested in review.

I did the last few edits on top of the original commit to make further review easier.

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -214,12 +214,12 @@ def get_prep_value(self, value):
                     obj = GDALRaster(obj)
                 except GDALException:
-                    raise ValueError("Couldn't create spatial object from lookup value.")
+                    raise ValueError("Couldn't create spatial object from lookup value \"%s\"." % obj)
```
We typically use single quotes for either the inner or outer string to avoid backslashes.

timgraham commented:
Okay, I guess the change in migrations is okay. It may very well be correct -- I don't have much time or interest in understanding it right now.

claudep commented:
Thanks to both for the patch and the review. I have nothing to comment on. I am also a newbie with raster and PostGIS. Daniel has already shown he's keen to address follow-up issues regarding the additions he brought, so I'm confident to let this in.

timgraham commented:
merged in bbfad84dd980a97174c3b061a3d1b5f1373c380d, thanks!

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -157,6 +160,82 @@ def get_placeholder(self, value, compiler, connection):
+        geometry or raster value properly and preserves any other lookup
+        parameters.
+        """
+        from django.contrib.gis.gdal import GDALRaster
```
Looks like the import here and in `GeometryField.get_prep_value()` introduced a GDAL dependency -- running the test suite without it (discovered on MySQL) results in `ImportError: cannot import name 'GDALRaster'`). Daniel, can you investigate?

yellowcap replied:
Here is a possible solution for the problem: https://github.com/django/django/pull/6654

I was not yet able to test this on a system without GDAL though. If its not too much trouble, could you try it out and see if you still get the error? Otherwise I'll set up a system to try it over the next few days.
